refactor: item name gathering script for equipment selection UIs

jhillacre opened the pull request:
This PR updates item name marshalling for equipment selection interfaces by refactoring `scr_weapons_equip` into `scr_get_item_names`, enhancing code clarity through explicit parameterization.

- **Refactored function**: Now takes explicit parameters for equipment type, subtype, unit type, and filtering options.
- **Replaces implicit dependencies**: Uses direct inputs rather than relying on different object contexts and globals.
- **Enums for improved readability**: Replaces ambiguous numeric codes with clear enums:
  - `eEQUIPMENT_TYPE` for equipment types.
  - `eEQUIPMENT_SUBTYPE` for equipment subtypes (Ranged/Melee).
  - `eUNIT_TYPE` for unit types (e.g., Infantry, Dreadnought).

jhillacre committed:
refactor: replace `scr_weapons_equip` with `scr_get_item_names` for clearer parameter-based item selection

jhillacre committed:
refactor(item_names): standardize item name loops.

- Migrate "(None)" and "(any)" to global constants for consistency across interfaces.
- Implement zero-based indexing for `item_name` to prevent out-of-bounds errors.
- Centralize slot-specific naming using `get_slot_name`.
- Convert `get_item_names` locals to underscore prefixed.

jhillacre commented:
- Migrated `"(None)"` and `"(any)"` to global constants (`global.item_name_none` and `global.item_name_any`) for consistent usage across interfaces.
- Introduced `get_slot_name` function to dynamically determine slot names based on `unit_type` and `equipment_type`, similar to `scr_get_item_names`.
- Switched to underscore prefixed locals in `scr_get_item_names` functions.

OH296 commented:
I'll go through this later in more detail but just as a heads up if your gonna start messing with the actual weapons struct as part of your work here make sure your synced with #64, as moves the sruct to a different file from the equipment values

OH296 commented on scripts/scr_get_item_names/scr_get_item_names.gml:
```diff
@@ -0,0 +1,646 @@
+        array_push(_item_names, "Lucifer Pattern Engine");
+    }
+}
+
```
I may be being dim here but should there not be a return statement for this set of functions? i think we still have Copy on Write behaviour enabled due to the age of this project https://manual.gamemaker.io/lts/en/GameMaker_Language/GML_Overview/Arrays.htm#:~:text=Arrays%20as%20Function%20Arguments,the%20original%20arrays%20as%20well.

jhillacre replied:
not dim, I thought arrays were by reference. I'll update.

OH296 replied:
yeah it's annoying but current situation is that only structs are by reference.

jhillacre replied:
updated in [13f5ab4](https://github.com/EttyKitty/ChapterMaster/pull/68/commits/13f5ab4adad09eb1fa446c26e25e4796582b6a0c)

OH296 commented on objects/obj_mass_equip/Draw_0.gml:
```diff
@@ -60,122 +60,145 @@ if (total_role_number>0){
+                        case 2: obj_ini.wep2[100, role] = buh; break;
+                        case 3:
+                            obj_ini.armour[100, role] = buh;
+                            if (buh == "Terminator Armour") obj_ini.mobi[100, role] = "";
```
somehow this needs to be moved to use equipment tags, may be outside of pr scope though
```suggestion
                            //TODO move to use equipment tags and look for "terminator" tag
                            if (buh == "Terminator Armour") obj_ini.mobi[100, role] = "";
```

OH296 reviewed (commented):
I haven't personally tested yet but have one question regarding Copy on Write and array functions

jhillacre committed:
chore: add todo comment

Co-authored-by: Nelsonh <81228864+OH296@users.noreply.github.com>

jhillacre committed:
fix: copy-on-write makes arrays not by ref.

jhillacre commented on scripts/scr_get_item_names/scr_get_item_names.gml:
```diff
@@ -0,0 +1,646 @@
+        case eUNIT_TYPE.Whirlwind:
+            _item_names = get_none_or_any_item_names(true, false);
+            switch (_equipment_type) {
+                case eEQUIPMENT_TYPE.PrimaryWeapon: push_whirlwind_missiles_item_names(_item_names); break;
+                case eEQUIPMENT_TYPE.SecondaryWeapon: push_whirlwind_pintle_item_names(_item_names); break;
```
Is this a desired change in functionality, or a place your saying I'm difference then the original?

If a desired change, should whirlwind missiles go somewhere?

I was reading the original as different lists for these two slots on whirlwinds.
https://github.com/EttyKitty/ChapterMaster/blob/5974efc152a65491f03a1b35e31d8d63840a2bbb/scripts/scr_weapons_equip/scr_weapons_equip.gml#L295-L307

OH296 commented on scripts/scr_get_item_names/scr_get_item_names.gml:
```diff
@@ -0,0 +1,646 @@
+            _item_names = get_none_or_any_item_names(true, false);
+            switch (_equipment_type) {
+                case eEQUIPMENT_TYPE.PrimaryWeapon: push_whirlwind_missiles_item_names(_item_names); break;
+                case eEQUIPMENT_TYPE.SecondaryWeapon: push_whirlwind_pintle_item_names(_item_names); break;
```
Sorry. ignore this.  i'm at fault, i completely misread the variables

jhillacre replied:
no worries, the old code needed updating for a reason.

jhillacre commented:
> I'll go through this later in more detail but just as a heads up if your gonna start messing with the actual weapons struct as part of your work here make sure your synced with #64, as moves the sruct to a different file from the equipment values

I'll read that PR. I need to find my next change out of the way of current work.

jhillacre commented:
Doing testing on this.

jhillacre committed:
fix: syntax, brackets, valid struct keys...

jhillacre commented:
Got a weird issue where "(None)" is always selected editing equipment on a marine, on all 5 slots. somehow `target_role` is always zero. `target_role` is only mentioned in my diff once, when deciding which item to check. I only reformatted the line for whitespace. 🤦

EttyKitty commented:
> somehow `target_role` is always zero

Funny arrays perhaps?

EttyKitty commented:
> * Migrated `"(None)"` and `"(any)"` to global constants (`global.item_name_none` and `global.item_name_any`) for consistent usage across interfaces.

You may want to use `macro` constants instead. Globals are not constants and are around 100 times more performance expensive than enums and maros (iirc), and generally should only be used in cases where you need a global scope variable, changeable after compile (not a constant). I think, that is.

jhillacre committed:
refactor: macro constant for none/any str

jhillacre committed:
fix: get selected item by name instead of ???

jhillacre committed:
fix: avoid predictable array resizing cascades

jhillacre committed:
fix: set  sentinel to -1 due to 0-based item index

jhillacre committed:
fix: skip quantity checks for "None" and "Any" items

jhillacre commented:
Fixed issues with None not being selectable and Any incorrectly triggering a "not enough" message. Any still appears non-functional, but this PR doesn’t aim to define its behaviour, only to refactor how selection and quantity checks handle it. Switched to macros for string constants.

EttyKitty commented on fonts/fnt_menu/fnt_menu.old.png:
You may want to revert edits to fnt files, as these are probably leftovers.

jhillacre replied:
Will do.

jhillacre replied:
Reverted in 8a1ddd6

EttyKitty commented on scripts/scr_get_item_names/scr_get_item_names.gml:
```diff
@@ -0,0 +1,806 @@
+// slot names differ by unit type then equipment type
+// HACK: GM doesn't like strings that look like numbers as struct keys
+//  so we prefix them with an underscore
+global.slot_names = {
```
If this variable is not intended to change at run-time, is it impossible to use an enum?
Or, considering that atm it's used only in one function, contain it locally in that function?

I just try to avoid global variables as much as possible, as I've seen notes that they are very bad practice from a performance standpoint. In this case it may be okay, but if it's possible to avoid, then better to avoid.

jhillacre replied:
These are not intended to change at run time.

I'm not sure what structure you should use in gml to have a constant lookup of strings based on two enums. Applying my python brain, I'd make a global dictionary with two levels and call it a day. In javascript, I'd make it a local in the outer closure of the function and access it inside.

Should I convert it to a switch inside the lookup function and delete the global?

EttyKitty replied:
>  Applying my python brain, I'd make a global dictionary with two levels and call it a day. In javascript, I'd make it a local in the outer closure of the function and access it inside.

I'm not sure if I'm misremembering, but I've seen people saying that you should only use `globals` when there is no way to get away without them and not turn your code unreadable/inefficient.
Or/and when you're referencing them rarely and are not doing any operations on them. So, if it wasn't something that runs every frame, maybe I would've been more okay with this.

> Should I convert it to a switch inside the lookup function and delete the global?

I'll suggest a couple of options, that I think may be viable, then you can decide on your own. I just can't explain it with words, as I need to fiddle with code.

EttyKitty replied:
Well, yeah. I came to this, which is your suggestion with the switch.
I'm not sure if "eEQUIPMENT_TYPE" enum is nessessary, because slots are just slots. First, second, etc.
But this is just how I think.
```suggestion
/// @description This function returns the name of the slot for a given unit type and equipment type.
/// @param {real} _unit_type - The type of unit to equip, see eUNIT_TYPE.
/// @param {real} _equipment_slot - The equipment slot number.
/// @returns {string} The name of the slot.
function get_slot_name(_unit_type, _equipment_slot) {
    // Map unit types to slot name maps
    switch (_unit_type) {
        case eUNIT_TYPE.Infantry:
            switch (_equipment_slot) {
                case 1: return "Left Hand";
                case 2: return "Right Hand";
                case 3: return "Armour";
                case 4: return "Gear";
                case 5: return "Mobility";
                default: return "Unknown";
            }
        case eUNIT_TYPE.Dreadnought:
            switch (_equipment_slot) {
                case 1: return "Left Arm";
                case 2: return "Right Arm";
                case 5: return "Accessory";
                default: return "Unknown";
            }
        case eUNIT_TYPE.LandRaider:
            switch (_equipment_slot) {
                case 1: return "Front";
                case 2: return "Sponson";
                case 3: return "Pintle";
                case 4: return "Upgrade";
                case 5: return "Accessory";
                default: return "Unknown";
            }
        case eUNIT_TYPE.Rhino:
            switch (_equipment_slot) {
                case 1: return "Weapon";
                case 4: return "Upgrade";
                case 5: return "Accessory";
                default: return "Unknown";
            }
        case eUNIT_TYPE.Predator:
            switch (_equipment_slot) {
                case 1: return "Turret";
                case 2: return "Sponsons";
                case 3: return "Pintle";
                case 4: return "Upgrade";
                case 5: return "Accessory";
                default: return "Unknown";
            }
        case eUNIT_TYPE.LandSpeeder:
            switch (_equipment_slot) {
                case 1: return "Primary";
                case 2: return "Secondary";
                default: return "Unknown";
            }
        case eUNIT_TYPE.Whirlwind:
            switch (_equipment_slot) {
                case 1: return "Missiles";
                case 2: return "Pintle";
                case 4: return "Upgrade";
                case 5: return "Accessory";
                default: return "Unknown";
            }
        default:
            return "Unknown";
    }
    return "Unknown";
}
```

EttyKitty replied:
There are probably some interesting ways of doing this with ds_ stuff, but I have 0 experience with them and just ignore this part of GM, because they are never garbage collected and have to be perfectly coded.

jhillacre replied:
I switched to using a switch in `get_slot_name` in ee82b82d.

EttyKitty commented on scripts/scr_get_item_names/scr_get_item_names.gml:
```diff
@@ -0,0 +1,806 @@
+    _item_names[index++] = "Sniper Rifle";
+    _item_names[index++] = "Storm Bolter";
+    _item_names[index++] = "Webber"; // 23
+    // future: not needed when we stop using GML's deprecated Copy-on-Write behavior
```
I may be misunderstanding, but can't you use the `@` accessor? It enables the pointer behaviour case-per-case.

jhillacre replied:
Ah, I'm new to gml, and it's javascript-looking gets me to assume I know what I'm doing. I'm wasn't aware of the `@` accessor. I'll update to use it.

jhillacre replied:
I made this change in cf8aefae

EttyKitty commented on scripts/scr_get_item_names/scr_get_item_names.gml:
```diff
@@ -0,0 +1,806 @@
+/// @description This function returns the hard-coded list of melee weapons.
+/// @param {array} _item_names - The list to append to.
+/// @returns {array} The appended list.
+function push_marine_melee_weapons_item_names(_item_names) {
+    var item_count = 15;
+    var initial_size = array_length(_item_names);
+    array_resize(_item_names, initial_size + item_count);
+
+    var index = initial_size;
```
Is there a reason to do these fixed size shenanigans and not just do `array_push`? I think it automatically runs the `array_resize` when needed.
I assume `array_push` is just more code lines, but you can use a loop, I think.

jhillacre replied:
These functions get called in Draw right now, I believe they are called every frame when the equip menu is open. I was thinking one resize at the the start would be cheaper than each time. I was also thinking that pushes were better than concat since there would be a left over array to garbage collect each frame. However, I'm not sure how the compiler works if it optimizes such things.

EttyKitty replied:
> These functions get called in Draw right now, I believe they are called every frame when the equip menu is open. I was thinking one resize at the the start would be cheaper than each time. I was also thinking that pushes were better than concat since there would be a left over array to garbage collect each frame. However, I'm not sure how the compiler works if it optimizes such things.

Yeah, this may be true. The fact that this runs every frame is dumb, but oh well. Out of the scope of this PR.
Using a single array, for loop and `array_push` is just more readable and less typo prone, imo.
I'll check this through a profiler. Maybe it's okay to leave it like this while it runs every frame.

EttyKitty replied:
> These functions get called in Draw right now, I believe they are called every frame when the equip menu is open.

Can you please double-check. I think they are called one time, when you click your mouse and open the menu. Then the return is stored in a variable and refreshed on other clicks.

jhillacre replied:
I confirmed that they are not spammed as I thought, thankfully.

Looking at cf8aefae, I touch all the call locations. 5 are in Draw_0.gml files, 1 in Step_0.gml and 1 is indirectly in draw via `scr_ui_settings`; that leaves 2 calls in a Mouse_50.gml. I did not look closer than noticing that these were called in draw events.

EttyKitty replied:
> I confirmed that they are not spammed as I thought, thankfully.

Then can you use a single array with item names and either a `for` loop or `array_foreach` with `array_push`, instead of these shenanigans?

EttyKitty replied:
Or `array_concat`, if you're not bothered by the fact that it returns a separate array and not modifies the original.

jhillacre replied:
In further testing, the calls in obj_mass_equip are per frame once a menu is opened.

EttyKitty commented on scripts/scr_get_item_names/scr_get_item_names.gml:
```diff
@@ -0,0 +1,806 @@
+                            _item_names = push_marine_melee_weapons_item_names(_item_names);
+                        }
+                    } else {
+                        show_error("scr_get_item_names: Invalid equipment subtype for infantry", true);
```
Is it intended to abort the game in case this fails?
I understand that this is a last resort scenario and don't suggest wrapping this into `try_and_report` or just try-catch, but is it impossible to just not abort it?
I assume if it gets bad, it'll crash anyway, but force crashing practice is a bit, eh.

jhillacre replied:
I assumed it was better to crash early and often. At my job, I want the developer who passed the wrong arguments to the function to know it went wrong, or for issues users encounter to get reported. In hindsight, I realize this doesn’t necessarily apply to CM if only one dev has that attitude.

What would you do in this situation?

EttyKitty replied:
> I want the developer who passed the wrong arguments to the function to know it went wrong, or for issues users encounter to get reported.
> 
> What would you do in this situation?

Well, this is true. I'm just afraid of these things slipping and then causing crashes, when they could've just caused an error popup. I think crash is just more nasty than an error, and I'm more in preference of soft error throwing and not intentional crashing.
If anything, I would prefer godot/unity way of wrapping all user code in the project into try-catch and continuing to run the game in all error cases, if the error is not critical. It just makes the game more robust, even if we're not around.

jhillacre replied:
In 3e60f571, I've switched to triggering an error popup. Let me know what you think.

EttyKitty replied:
> In [3e60f57](https://github.com/EttyKitty/ChapterMaster/commit/3e60f5713469dfff41df470120ea976b912826ec), I've switched to triggering an error popup. Let me know what you think.

Yeah, that works. I wanted to separate exception handling and error display as well, good idea.
But it'll probably now clash with https://github.com/EttyKitty/ChapterMaster/commit/b2a0d5015ce193dc40e72c7c2b77e36b64572de7.
So we're in for a fun conflict solving.

I can solve this conflicts myself at the end, if you would prefer.

jhillacre replied:
sure, you know what you were up to and reviewed what I was up to.

EttyKitty commented on scripts/scr_get_item_names/scr_get_item_names.gml:
```diff
@@ -0,0 +1,806 @@
+    }
+};
+
+/// @description This function returns the name of the slot for a given unit type and equipment type.
+/// @param {real} _unit_type - The type of unit to equip, see eUNIT_TYPE.
+/// @param {real} _equipment_type - The type of equipment to equip, see eEQUIPMENT_TYPE.
+/// @returns {string} The name of the slot.
+function get_slot_name(_unit_type, _equipment_type) {
+    var _unit_key_str = $"_{_unit_type}";
+    var _equip_key_str = $"_{_equipment_type}";
+
+    if (global.slot_names[_unit_key_str] && global.slot_names[_unit_key_str][_equip_key_str]) {
+        return global.slot_names[_unit_key_str][_equip_key_str];
+    }
+    return "Unknown";
+}
```
Well, yeah. I came to this, which is your suggestion with the switch.
I'm not sure if "eEQUIPMENT_TYPE" enum is nessessary, because slots are just slots. First, second, etc.
But this is just how I think.
```suggestion
/// @description This function returns the name of the slot for a given unit type and equipment type.
/// @param {real} _unit_type - The type of unit to equip, see eUNIT_TYPE.
/// @param {real} _equipment_slot - The equipment slot number.
/// @returns {string} The name of the slot.
function get_slot_name(_unit_type, _equipment_slot) {
    // Map unit types to slot name maps
    switch (_unit_type) {
        case eUNIT_TYPE.Infantry:
            switch (_equipment_slot) {
                case 1: return "Left Hand";
                case 2: return "Right Hand";
                case 3: return "Armour";
                case 4: return "Gear";
                case 5: return "Mobility";
                default: return "Unknown";
            }
        case eUNIT_TYPE.Dreadnought:
            switch (_equipment_slot) {
                case 1: return "Left Arm";
                case 2: return "Right Arm";
                case 5: return "Accessory";
                default: return "Unknown";
            }
        case eUNIT_TYPE.LandRaider:
            switch (_equipment_slot) {
                case 1: return "Front";
                case 2: return "Sponson";
                case 3: return "Pintle";
                case 4: return "Upgrade";
                case 5: return "Accessory";
                default: return "Unknown";
            }
        case eUNIT_TYPE.Rhino:
            switch (_equipment_slot) {
                case 1: return "Weapon";
                case 4: return "Upgrade";
                case 5: return "Accessory";
                default: return "Unknown";
            }
        case eUNIT_TYPE.Predator:
            switch (_equipment_slot) {
                case 1: return "Turret";
                case 2: return "Sponsons";
                case 3: return "Pintle";
                case 4: return "Upgrade";
                case 5: return "Accessory";
                default: return "Unknown";
            }
        case eUNIT_TYPE.LandSpeeder:
            switch (_equipment_slot) {
                case 1: return "Primary";
                case 2: return "Secondary";
                default: return "Unknown";
            }
        case eUNIT_TYPE.Whirlwind:
            switch (_equipment_slot) {
                case 1: return "Missiles";
                case 2: return "Pintle";
                case 4: return "Upgrade";
                case 5: return "Accessory";
                default: return "Unknown";
            }
        default:
            return "Unknown";
    }
    return "Unknown";
}
```

EttyKitty replied:
There are probably some interesting ways of doing this with `ds_` stuff, but I have 0 experience with them and avoid them, as they are not garbage collected and have to be perfectly coded.

EttyKitty commented on scripts/scr_get_item_names/scr_get_item_names.gml:
```diff
@@ -0,0 +1,806 @@
+        "_5": "Accessory"
+    },
+    "_51": { // Rhino
+        "_1": "Weapon",
+        "_4": "Upgrade",
+        "_5": "Accessory"
+    },
+    "_52": { // Predator
+        "_1": "Turret",
+        "_2": "Sponsons",
+        "_3": "Pintle",
+        "_4": "Upgrade",
+        "_5": "Accessory"
+    },
+    "_53": { // Land Speeder
+        "_1": "Primary",
+        "_2": "Secondary"
+    },
+    "_54": { // Whirlwind
+        "_1": "Missiles",
+        "_2": "Pintle",
+        "_4": "Upgrade",
+        "_5": "Accessory"
+    }
+};
+
+/// @description This function returns the name of the slot for a given unit type and equipment type.
+/// @param {real} _unit_type - The type of unit to equip, see eUNIT_TYPE.
+/// @param {real} _equipment_type - The type of equipment to equip, see eEQUIPMENT_TYPE.
+/// @returns {string} The name of the slot.
+function get_slot_name(_unit_type, _equipment_type) {
+    var _unit_key_str = $"_{_unit_type}";
+    var _equip_key_str = $"_{_equipment_type}";
+
+    if (global.slot_names[_unit_key_str] && global.slot_names[_unit_key_str][_equip_key_str]) {
+        return global.slot_names[_unit_key_str][_equip_key_str];
+    }
+    return "Unknown";
+}
+
```
Well, yeah. I came to this, which is your suggestion with the switch.
I'm not sure if "eEQUIPMENT_TYPE" enum is necessary, because slots are just slots. First, second, etc. I don't think that they appear as magic numbers in this case.
But this is just how I think.
```suggestion


/// @description This function returns the name of the slot for a given unit type and equipment type.
/// @param {real} _unit_type - The type of unit to equip, see eUNIT_TYPE.
/// @param {real} _equipment_slot - The equipment slot number.
/// @returns {string} The name of the slot.
function get_slot_name(_unit_type, _equipment_slot) {
    // Map unit types to slot name maps
    switch (_unit_type) {
        case eUNIT_TYPE.Infantry:
            switch (_equipment_slot) {
                case 1: return "Left Hand";
                case 2: return "Right Hand";
                case 3: return "Armour";
                case 4: return "Gear";
                case 5: return "Mobility";
                default: return "Unknown";
            }
        case eUNIT_TYPE.Dreadnought:
            switch (_equipment_slot) {
                case 1: return "Left Arm";
                case 2: return "Right Arm";
                case 5: return "Accessory";
                default: return "Unknown";
            }
        case eUNIT_TYPE.LandRaider:
            switch (_equipment_slot) {
                case 1: return "Front";
                case 2: return "Sponson";
                case 3: return "Pintle";
                case 4: return "Upgrade";
                case 5: return "Accessory";
                default: return "Unknown";
            }
        case eUNIT_TYPE.Rhino:
            switch (_equipment_slot) {
                case 1: return "Weapon";
                case 4: return "Upgrade";
                case 5: return "Accessory";
                default: return "Unknown";
            }
        case eUNIT_TYPE.Predator:
            switch (_equipment_slot) {
                case 1: return "Turret";
                case 2: return "Sponsons";
                case 3: return "Pintle";
                case 4: return "Upgrade";
                case 5: return "Accessory";
                default: return "Unknown";
            }
        case eUNIT_TYPE.LandSpeeder:
            switch (_equipment_slot) {
                case 1: return "Primary";
                case 2: return "Secondary";
                default: return "Unknown";
            }
        case eUNIT_TYPE.Whirlwind:
            switch (_equipment_slot) {
                case 1: return "Missiles";
                case 2: return "Pintle";
                case 4: return "Upgrade";
                case 5: return "Accessory";
                default: return "Unknown";
            }
        default:
            return "Unknown";
    }
    return "Unknown";
}

```

jhillacre replied:
I switched to a switch statement in `get_slot_name` in ee82b82d, but I’d like to keep the `eEQUIPMENT_TYPE` enum. Even in simple cases, using enums can set a good example of avoiding magic numbers, especially for future readers who may appreciate the extra clarity in the case statements.

EttyKitty replied:
>  Even in simple cases, using enums can set a good example of avoiding magic numbers, especially for future readers who may appreciate the extra clarity in the case statements.

But they are also confusing, imo, with how they are named.
If we use more different vehicles/unit types, that use these slots for different things, how are we going to name enums? `eEQUIPMENT_TYPE.MobilityAccessorySupportBackpackLightsCristmasTree`?
And also, what do I gather from the name of the enum? Is it a Mobility slot or an Accessory slot after all? Which one in what case?

And it's not an `eEQUIPMENT_TYPE/_equipment_type`, it's just a slot number.
You're differentiating them across different unit/vehicles types in this exact function. Before that, they are just slot numbers.

It's just a bad precedent, where, imo, we're just trying to replace all numbers with enums because we can. It brings no clarity and is over the top in this case. Very hard IMO, in this case.

And, also IMO, not every number is a magic number and shouldn be replaced with an enum.

EttyKitty replied:
If you really hate numbers, the only semi-beneficial enum in this case, imo, is:
```
enum eEQUIPMENT_SLOT {
    First = 1,
    Second = 2,
    Third = 3,
    Fourth = 4,
    Fifth = 5
}
or
enum eEQUIPMENT_SLOT {
    One = 1,
    Two = 2,
    Three = 3,
    Four = 4,
    Five = 5
}
```

EttyKitty commented on scripts/scr_get_item_names/scr_get_item_names.gml:
```diff
@@ -0,0 +1,806 @@
+function scr_get_item_names(_equipment_type, _equipment_subtype, _unit_type, _include_company_standard=false, _show_available_only=false, _master_crafted_only=false) {
+    show_debug_message("scr_get_item_names: " + string(_equipment_type) + ", " + string(_equipment_subtype) + ", " + string(_unit_type) + ", " + string(_include_company_standard) + ", " + string(_show_available_only) + ", " + string(_master_crafted_only));
+    var _item_names;
+    switch(_unit_type) {
```
I think you may want to do this:
```var _item_names = [];```
otherwise there is a crash if you click on a slot that should have no items, caused by this line:
`show_debug_message("scr_get_item_names: " + string(_item_names));`

jhillacre replied:
Not sure why GitHub didn't mark this as outdated, but I've removed the local _item_names and the debug message, so this isn't an issue in the current PR.

jhillacre committed:
revert: unintentional changes.

jhillacre committed:
refactor: remove redundant array assignment in scr_get_item_names calls

jhillacre committed:
refactor(error-handling): centralize error display and logging via show_error_popup()

- Added assert_error_popup to handle display assertion failures.
- Updated array_to_string_list to filter unwanted values from stack traces.

jhillacre committed:
fix: return values in assert popup because then it is useful, remove debug logging.

jhillacre committed:
refactor: replace global slot names with switch statement in get_slot_name()

EttyKitty commented on scripts/scr_array_functions/scr_array_functions.gml:
```diff
@@ -80,7 +80,10 @@ function array_to_string_list(_array) {
     }
     for (var i = 0; i < array_length(_array); i++) {
-        _string_list += _array[i];
+		if (_array[i] == 0) { // top of a stack trace from debug_get_callstack
+			continue;
+		}
+        _string_list += string(_array[i]);
```
Are you sure that you want to do this here? I'm concerned that if you run this function unrelated to `debug_get_callstack`, expecting a 0 in the list as well, this will cause it to skip it.
Maybe do this in the error throw function instead, pre-bake the array before this function.

jhillacre replied:
moved in [0cbca70](https://github.com/EttyKitty/ChapterMaster/pull/68/commits/0cbca7097e52c19650666e726c4d6b108d802e1b)

jhillacre committed:
feat(equipment): unify weapon selection display with multi-column layout

jhillacre committed:
fix: let array_to_string_list show zeros.

jhillacre commented:
Addressed issues in mass equipping via `scr_get_item_names` and `get_slot_names` involving unhandled roles. Commit 81ad08fa expands `eUNIT_TYPE` to include these roles and updates relevant switch cases. Refactored `obj_mass_equip` to consolidate item lookup and drawing into a single loop, maintaining the assumption that items are equippable in either hand. This refactoring was a side-effect of figuring out the arguments for the above functions.

EttyKitty commented:
> expands `eUNIT_TYPE` to include these roles and updates relevant switch cases.

Are you sure you want to do this? You're efficiently duplicating the `Role` enum from `scr_initialize_custom`. Minus the fact that it uses 1 for chapter master and doesn't support vehicles.

Also, not sure if you saw [this](https://github.com/EttyKitty/ChapterMaster/pull/68#discussion_r1835307119) or if github hid it before you were able to.

jhillacre commented:
I do be hating magical numbers. It feels like each instance is 5-45 minutes of life gone to understand it.

> Are you sure you want to do this? You're efficiently duplicating the Role enum from scr_initialize_custom. Minus the fact that it uses 1 for chapter master and doesn't support vehicles.

`scr_get_item_name` deals with vehicles. `enum Role` is a partial list, and so is the new one. It doesn't have to be perfect now, the enums can be merged for another refactor PR, and get guaranteed to be complete. I'll rename my `1` to chapter master since that is probably more true than just infantry, considering the list.

> Also, not sure if you saw [this](https://github.com/EttyKitty/ChapterMaster/pull/68#discussion_r1835307119) or if github hid it before you were able to.

If the variables regarding it where all called slot, that would go a long way, and avoid the enum naming issue you brought up. That's beyond the scope of what I wanted to do here, so I'll remove the `eEQUIPMENT_TYPE` enum.

jhillacre committed:
wip: combined Role enums.

jhillacre committed:
refactor: revert equip type enum to slot #, rename subtype enum, semantic reordering of get_item_names args

jhillacre commented:
@EttyKitty 
Reverted eEQUIPMENT_TYPE back to slot numbers (1–5), as we talked about, and clarified their use in context as slots numbers. While continuing, integrating the role enum felt manageable without expanding the PR scope too much, so I went ahead with your suggestion on that. I recased eRole -> eROLE for the new code style, and updated the entries. Also, I reordered `scr_get_item_names` arguments for clarity.

jhillacre commented:
The conflicts weren't too crazy, I rebased to resolve the pair of conflicts.

jhillacre committed:
fix: old enum names in asserts, use string interp.

jhillacre committed:
fix: caps matter.

EttyKitty commented:
> While continuing, integrating the role enum felt manageable without expanding the PR scope too much, so I went ahead with your suggestion on that.

Btw, I missed it, but why you decided to add roles into the original enum in the first place?

I assume, to use it as an index to reference an array directly.

EttyKitty reviewed (commented):
I mostly happy with this PR and think it should be merged at this point.
Everything else is just my nitpicking and can be addressed later, if still needed.

Just need to bring it up-to-date with the release branch and see if it works properly and doesn't crash.

jhillacre commented:
> > While continuing, integrating the role enum felt manageable without expanding the PR scope too much, so I went ahead with your suggestion on that.
> 
> Btw, I missed it, but why you decided to add roles into the original enum in the first place?
> 
> I assume, to use it as an index to reference an array directly.

Initially, I made a separate enum out of not knowing about the original. After you pointed it out, I looked into it more and decided to extend the existing roles enum instead, combining marine and vehicle roles in one enum. I wanted the role argument to have a single enum type, and the existing numbers don’t overlap on purpose. Spotted this section that could use roles enum, with potential to use the vehicle entries: https://github.com/EttyKitty/ChapterMaster/blob/18c76daa57710e7c977d8a59cb1faf382de38325/objects/obj_controller/Create_0.gml#L1563-L1591

jhillacre committed:
chore: arg name update.

EttyKitty commented:
> Spotted this section that could use roles enum, with potential to use the vehicle entries:

There are a lot of sections like this. Not even with vehicles, but marine roles that still use indices and not enums.

jhillacre committed:
fix: hand slot tabs not switching between melee and ranged / relic or not.

jhillacre commented:
Apologies for toggling the draft status again.

Fixed the last couple of bugs, and there are no crashes now. You can mass equip via chapter settings, bulk equips via selection, and equip infantry and vehicles in each slot. Noted that the chapter creation screen for pre-setting chapter settings doesn’t work; nothing happens on click, but this was also broken on the release branch.

EttyKitty commented:
> Noted that the chapter creation screen for pre-setting chapter settings doesn’t work; nothing happens on click, but this was also broken on the release branch.

Funny, but for me, it actually crashes, due to typo. Not related to this PR though.
*(okay, this got solved after checking out on your force pushed head)

A note: can you perhaps consider changing the name of marine weapon slots to “First Weapon” and “Second Weapon” or something along the lines? Just not right/left, and not primary/secondary (as that's decided in the code, independent of slots).

The problem is that these are not left/right-hand slots, as we don't have a properly working dual-welding mechanic, and they just function as two weapons. Sometimes they can be dual-wielded, but most of the time are just switched between, depending on the combat type.
So imo, it ends up a bit misleading and makes you expect dual-wielding, when it's just not the case.

EttyKitty commented on objects/obj_creation_popup/Draw_0.gml:
```diff
@@ -110,59 +110,70 @@ if (col_shift){
+                        )
+                    );
+
+                    if (!is_invalid) {
+                        obj_creation.cooldown = 8000;
+
+                        item_name = [];
+                        scr_get_item_names(
+                            item_name,
+                            unit_type, // eROLE
+                            gg, // slot
+                            eENGAGEMENT.Ranged,
+                            false,  // no company standard
+                            false   // don't limit to available items
+                        );
                     }
```
I think this should solve the creation screen role equip issue.
```suggestion
                    if (!is_invalid) {
                        obj_creation.cooldown = 8000;
                        tab=1;
                        target_gear=gg;
                        item_name = [];
                        scr_get_item_names(
                            item_name,
                            unit_type, // eROLE
                            gg, // slot
                            eENGAGEMENT.Ranged,
                            false,  // no company standard
                            false   // don't limit to available items
                        );
                    }
```

jhillacre replied:
this fixes it. 👍

jhillacre committed:
fix: creation equip tab item list not showing.

Co-authored-by: EttyKitty <20323032+EttyKitty@users.noreply.github.com>

jhillacre committed:
content: rename slots to not imply handedness of slots.

jhillacre commented:
> > Noted that the chapter creation screen for pre-setting chapter settings doesn’t work; nothing happens on click, but this was also broken on the release branch.
> 
> Funny, but for me, it actually crashes, due to typo. Not related to this PR though. *(okay, this got solved after checking out on your force pushed head)
> 
> A note: can you perhaps consider changing the name of marine weapon slots to “First Weapon” and “Second Weapon” or something along the lines? Just not right/left, and not primary/secondary (as that's decided in the code, independent of slots).
> 
> The problem is that these are not left/right-hand slots, as we don't have a properly working dual-welding mechanic, and they just function as two weapons. Sometimes they can be dual-wielded, but most of the time are just switched between, depending on the combat type. So imo, it ends up a bit misleading and makes you expect dual-wielding, when it's just not the case.

Got it. Changed in 1662812

OH296 commented:
> > Noted that the chapter creation screen for pre-setting chapter settings doesn’t work; nothing happens on click, but this was also broken on the release branch.
> 
> Funny, but for me, it actually crashes, due to typo. Not related to this PR though. *(okay, this got solved after checking out on your force pushed head)
> 
> A note: can you perhaps consider changing the name of marine weapon slots to “First Weapon” and “Second Weapon” or something along the lines? Just not right/left, and not primary/secondary (as that's decided in the code, independent of slots).
> 
> The problem is that these are not left/right-hand slots, as we don't have a properly working dual-welding mechanic, and they just function as two weapons. Sometimes they can be dual-wielded, but most of the time are just switched between, depending on the combat type. So imo, it ends up a bit misleading and makes you expect dual-wielding, when it's just not the case.

sorry i realise i'm very late in the hour in on this, but i find this more misleading, each slot refers to the left and the right side for the purpose of drawing. Taking away the context of left and right now seems misleading imo that is

EttyKitty commented:
> but i find this more misleading, each slot refers to the left and the right side for the purpose of drawing. Taking away the context of left and right now seems misleading imo that is

Drawing is not really as consequential to make slot names misleading. You just do it once and remember which side is responsible for what slot.

Dual-wielding implications are, on the other side, misleading and consequential enough.
You can equip a heavy bolter into your right hand and a combat knife into the left one. What does this mean? Nothing. The other weapon won't be even drawn.
You can equip a bolter and a bolt pistol into each hand. They'll be drawn separately but will do nothing in terms of gameplay.

So, basically, aside from drawing sides, these don't mean anything. And drawing side imo is not as important and not as misleading.

You're not even misleading per se. You're just not telling what slot corresponds to what drawing side.

jhillacre commented:
let's merge. With this slot naming centralization and getting more places to uses it, switching slots naming to be more thematic or semantic will be easier later.
